sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -671,7 +674,7 @@ case class AlterTableRecoverPartitionsCommand(
         val value = ExternalCatalogUtils.unescapePathName(ps(1))
         if (resolver(columnName, partitionNames.head)) {
           scanPartitions(spark, fs, filter, st.getPath, spec ++ Map(partitionNames.head -> value),
-            partitionNames.drop(1), threshold, resolver)
+            partitionNames.drop(1), threshold, resolver, listFilesInParallel = false)
```

Comment:
This change might introduce performance regression. Do you know why it works when using `.par` previously?

Reply:
@MaxGekk Do you have a stack trace on each thread when a dead lock occurs?

Reply:
Thank you attaching the stack trace. I have just looked at it. It looks strange to me. Every thread is `waiting for`. No blocker is there, only one `locked` exists.
In typical case, a deadlock occurs due to existence of blocker as attached stack trace in #22221

I will investigate it furthermore tomorrow if we need to use this implementation instead of reverting it to the original implementation to use Scala parallel collection.

```
...
        - parking to wait for  <0x0000000793c0d610> (a scala.concurrent.impl.Promise$CompletionLatch)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
        at scala.concurrent.impl.Promise$DefaultPromise.tryAwait(Promise.scala:206)
        at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:222)
        at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:227)
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:220)
        at org.apache.spark.util.ThreadUtils$.parmap(ThreadUtils.scala:317)
        at org.apache.spark.sql.execution.command.AlterTableRecoverPartitionsCommand.scanPartitions(ddl.scala:690)
        at org.apache.spark.sql.execution.command.AlterTableRecoverPartitionsCommand.run(ddl.scala:626)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
        - locked <0x0000000793b04e88> (a org.apache.spark.sql.execution.command.ExecutedCommandExec)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
...
```

Reply:
@kiszk Thanks for the investigation! Please take a look at the root cause? If unable to figure it out, we need to revert it back to `.par`. Thanks!

Reply:
cc @zsxwing had a few offline comments about the original PR for `parmap`. He will post them soon.

Reply:
@MaxGekk can we remove this `parmap` overload? It's pretty easy to cause deadlock. The `parmap` overload without the `ec` parameter is fine since it doesn't need a user specified thread pool.

Reply:
> The parmap overload without the ec parameter is fine ...

I didn't use it because each recursive call creates new fork join pool. This can create significant amount of threads - `n` ^ `d` threads where `n` - max size of fork join pool, `d` - average depth.

Reply:
@MaxGekk could you revert to use Scala `par`?

Reply:
Does it mean there is no avaiable thread in a given thread pool when a problem try to execute a new `Future`?

Reply:
@kiszk Right, all `Future`s do the same - trying to execute another `Future` on the same fixed thread pool.

Reply:
I am worry whether the similar deadlock may occur in other places due to
* larger parallelism than the fixed thread pool
* nested parallelism like this

I also realized there is another `parmap` implementation uses thread pool. Can we use  another implemetation?

Reply:
> Can we use another implemetation?

This is what @zsxwing proposed. Please, look at my comment https://github.com/apache/spark/pull/22233#discussion_r213075357

Reply:
Got it. sorry for my overlooking.

Are other places safe where parallelism would not reach the fixed thread pool size?